File: swift/Sources/CircleAI/Hosting.swift
// Hosting.swift
//
// IAIObserver + AIOptions.

import Foundation

public protocol IAIObserver: Sendable {
    func onStarted() async
    func onStopped() async
    func onChatCompleted(_ response: ChatResponse) async
    func onStreamStarted(_ modelId: String) async
    func onStreamCompleted(_ modelId: String, tokenCount: Int) async
    func onToolInvoked(_ toolName: String, success: Bool) async
    func onModelFetching(_ modelId: String, autoSelected: Bool) async
    func onUpgradeAvailable(_ upgrade: UpgradeInfo) async
}

/// No-op base. Conform and override only what you care about.
open class AIObserverBase: IAIObserver, @unchecked Sendable {
    public init() {}
    open func onStarted() async {}
    open func onStopped() async {}
    open func onChatCompleted(_ response: ChatResponse) async {}
    open func onStreamStarted(_ modelId: String) async {}
    open func onStreamCompleted(_ modelId: String, tokenCount: Int) async {}
    open func onToolInvoked(_ toolName: String, success: Bool) async {}
    open func onModelFetching(_ modelId: String, autoSelected: Bool) async {}
    open func onUpgradeAvailable(_ upgrade: UpgradeInfo) async {}
}

/// Configuration bag for the long-lived butler service. All fields have safe
/// defaults so callers can `AIOptions()` and get a working instance. Ported
/// from `AIOptions` (v2/v3 surface: sensorium, RAG, persona, feedback, agentic,
/// loopback, cloud fallback, thermal, scheduled tasks, affect, goals).
///
/// `@unchecked Sendable`: it holds reference-typed collaborators (tool bridge,
/// stores) that aren't themselves `Sendable`; the options bag is treated as an
/// immutable configuration snapshot, mirroring the C# init-only record.
public struct AIOptions: @unchecked Sendable {
    // Model
    public let modelId: String?
    public let modelPath: String?

    // Inference
    public let systemPrompt: String
    public let defaultGenerationOptions: GenerationOptions?
    public let contextSize: Int?
    public let threadCount: Int?
    public let warmOnStart: Bool

    // Tools
    public let toolBridge: (any IToolBridge)?

    // Observers
    public let observer: (any IAIObserver)?

    // v2.0 — Sensorium
    public let deviceContext: (any IDeviceContext)?
    public let catalogClient: ModelScopeCatalogClient?
    public let requiredCapabilities: ChatCapability
    public let checkForUpgradesOnStart: Bool
    public let modelStorageDirectory: String?

    // v2.0 — Memory / RAG
    public let episodicMemory: (any IEpisodicMemoryStore)?
    public let ragBuilder: RagContextBuilder?
    public let ragTopK: Int

    // v2.0 — Persona evolution
    public let personaStore: (any IPersonaStore)?
    public let personaUserId: String

    // v2.0 — Feedback
    public let feedbackStore: (any IFeedbackStore)?

    // v2.0 — Agentic loop
    public let agenticMaxIterations: Int?

    // Loopback endpoint
    public let loopbackPort: Int
    public let loopbackToken: String?

    // v2.1 — Model management
    public let modelStorageDir: String
    public let wifiOnlyModelDownload: Bool

    // v2.1 — Cloud fallback
    public let cloudFallbackEnabled: Bool
    public let cloudFallbackToken: String?
    public let cloudFallbackRamThresholdBytes: Int64

    // v3.0 — Scheduled tasks
    public let scheduledTaskStore: (any IScheduledTaskStore)?

    // v3.0 — Affect
    public let affectStore: (any IAffectStore)?

    // v3.0 — Goals
    public let goalStore: (any IGoalStore)?

    public init(
        modelId: String? = nil,
        modelPath: String? = nil,
        systemPrompt: String = "You are B!, a helpful on-device assistant.",
        defaultGenerationOptions: GenerationOptions? = nil,
        contextSize: Int? = nil,
        threadCount: Int? = nil,
        warmOnStart: Bool = true,
        toolBridge: (any IToolBridge)? = nil,
        observer: (any IAIObserver)? = nil,
        deviceContext: (any IDeviceContext)? = nil,
        catalogClient: ModelScopeCatalogClient? = nil,
        requiredCapabilities: ChatCapability = .defaultCap,
        checkForUpgradesOnStart: Bool = false,
        modelStorageDirectory: String? = nil,
        episodicMemory: (any IEpisodicMemoryStore)? = nil,
        ragBuilder: RagContextBuilder? = nil,
        ragTopK: Int = 5,
        personaStore: (any IPersonaStore)? = nil,
        personaUserId: String = "default",
        feedbackStore: (any IFeedbackStore)? = nil,
        agenticMaxIterations: Int? = nil,
        loopbackPort: Int = 0,
        loopbackToken: String? = nil,
        modelStorageDir: String? = nil,
        wifiOnlyModelDownload: Bool = true,
        cloudFallbackEnabled: Bool = false,
        cloudFallbackToken: String? = nil,
        cloudFallbackRamThresholdBytes: Int64 = 2 * 1024 * 1024 * 1024,
        scheduledTaskStore: (any IScheduledTaskStore)? = nil,
        affectStore: (any IAffectStore)? = nil,
        goalStore: (any IGoalStore)? = nil
    ) {
        self.modelId = modelId
        self.modelPath = modelPath
        self.systemPrompt = systemPrompt
        self.defaultGenerationOptions = defaultGenerationOptions
        self.contextSize = contextSize
        self.threadCount = threadCount
        self.warmOnStart = warmOnStart
        self.toolBridge = toolBridge
        self.observer = observer
        self.deviceContext = deviceContext
        self.catalogClient = catalogClient
        self.requiredCapabilities = requiredCapabilities
        self.checkForUpgradesOnStart = checkForUpgradesOnStart
        self.modelStorageDirectory = modelStorageDirectory
        self.episodicMemory = episodicMemory
        self.ragBuilder = ragBuilder
        self.ragTopK = ragTopK
        self.personaStore = personaStore
        self.personaUserId = personaUserId
        self.feedbackStore = feedbackStore
        self.agenticMaxIterations = agenticMaxIterations
        self.loopbackPort = loopbackPort
        self.loopbackToken = loopbackToken
        self.modelStorageDir = modelStorageDir
            ?? (URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
                .appendingPathComponent("models").path)
        self.wifiOnlyModelDownload = wifiOnlyModelDownload
        self.cloudFallbackEnabled = cloudFallbackEnabled
        self.cloudFallbackToken = cloudFallbackToken
        self.cloudFallbackRamThresholdBytes = cloudFallbackRamThresholdBytes
        self.scheduledTaskStore = scheduledTaskStore
        self.affectStore = affectStore
        self.goalStore = goalStore
    }

    /// Generates a cryptographically random 32-byte token, base64-encoded. Used
    /// by `HttpLoopbackEndpoint` when `loopbackToken` is nil. Mirrors
    /// `AIOptions.GenerateRandomToken`.
    public static func generateRandomToken() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        for i in 0..<bytes.count { bytes[i] = UInt8.random(in: 0...255) }
        return Data(bytes).base64EncodedString()
    }
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/hosting/Hosting.kt
// Hosting.kt
//
// Kotlin port of the CircleAI.Hosting observability + configuration surface —
// the C# reference is the EXACT spec (IAIObserver.cs, AIOptions.cs). Neutral
// observer hook with event records + the butler configuration bag.
//
// Cancellation is via coroutine cancellation, so the C# `CancellationToken`
// parameters are dropped. `ValueTask`/`Task` map to `suspend fun`; event
// `record`s map to `data class`. `Guid` -> UUID, `DateTimeOffset` -> Instant,
// `TimeSpan` -> Duration.

package com.bhengubv.circleai.hosting

import com.bhengubv.circleai.catalog.ModelScopeCatalogClient
import com.bhengubv.circleai.device.IDeviceContext
import com.bhengubv.circleai.inference.GenerationOptions
import com.bhengubv.circleai.memory.IAffectStore
import com.bhengubv.circleai.memory.IEpisodicMemoryStore
import com.bhengubv.circleai.memory.IFeedbackStore
import com.bhengubv.circleai.memory.IGoalStore
import com.bhengubv.circleai.memory.IPersonaStore
import com.bhengubv.circleai.models.ChatMessage
import com.bhengubv.circleai.models.UpgradeInfo
import com.bhengubv.circleai.selector.ChatCapability
import com.bhengubv.circleai.tools.IToolBridge
import com.bhengubv.circleai.tools.ToolInvocation
import com.bhengubv.circleai.tools.ToolResult
import java.security.SecureRandom
import java.time.Duration
import java.time.Instant
import java.util.Base64
import java.util.UUID

// ---------------------------------------------------------------------------
// Event records (IAIObserver.cs)
// ---------------------------------------------------------------------------

/**
 * Payload delivered to [IAIObserver.onChatCompletedAsync]. Carries the full
 * conversation and the model's reply. Mirrors C# `AIChatEvent`.
 */
data class AIChatEvent(
    val correlationId: UUID,
    val messages: List<ChatMessage>,
    val response: String,
    val elapsed: Duration,
    val timestamp: Instant,
)

/**
 * Payload delivered to [IAIObserver.onStreamStartedAsync] and
 * [IAIObserver.onStreamCompletedAsync]. Mirrors C# `AIStreamEvent`.
 */
data class AIStreamEvent(
    val correlationId: UUID,
    val messages: List<ChatMessage>,
    val elapsed: Duration,
    val tokenCount: Int,
    val timestamp: Instant,
)

/** Payload delivered to [IAIObserver.onToolInvokedAsync]. Mirrors C# `AIToolEvent`. */
data class AIToolEvent(
    val correlationId: UUID,
    val invocation: ToolInvocation,
    val result: ToolResult,
    val elapsed: Duration,
    val timestamp: Instant,
)

/**
 * (RT-04) Why a brownout swap fired. Sized so future causes can be added
 * without breaking ABI. Mirrors C# `BrownoutReason`.
 */
enum class BrownoutReason {
    /** OS-reported memory pressure. */
    MemoryPressure,
    /** Battery dropped below the brownout floor. */
    BatteryFloor,
    /** Thermal throttle declared the runtime must downshift. */
    ThermalCritical,
    /** Application requested the swap explicitly. */
    Manual,
}

// ---------------------------------------------------------------------------
// IAIObserver (IAIObserver.cs)
// ---------------------------------------------------------------------------

/**
 * Observability hook for the butler service. Receives lifecycle and inference
 * events. All methods are optional (default = no-op) and must complete quickly.
 * Mirrors C# `IAIObserver`.
 *
 * Thread safety: methods may be called concurrently. Implementations must be
 * thread-safe. Error isolation: exceptions thrown by observer methods are
 * caught by [AIService] and never propagate to the caller.
 */
interface IAIObserver {
    /** Called once after the model has loaded and Butler is ready. */
    suspend fun onStartedAsync() {}

    /** Called once when Butler is stopping / being disposed. */
    suspend fun onStoppedAsync() {}

    /** Called after a complete (non-streaming) chat response has been generated. */
    suspend fun onChatCompletedAsync(event: AIChatEvent) {}

    /** Called when a streaming response emits its first token (tokenCount == 0). */
    suspend fun onStreamStartedAsync(event: AIStreamEvent) {}

    /** Called after a streaming response has finished (all tokens yielded, or cancelled). */
    suspend fun onStreamCompletedAsync(event: AIStreamEvent) {}

    /** Called after a tool invocation has completed (success or failure). */
    suspend fun onToolInvokedAsync(event: AIToolEvent) {}

    /**
     * Called once when [AIService.startAsync] has resolved which model to load.
     * Fires before the actual file fetch/load so observers can surface progress UI.
     */
    suspend fun onModelFetchingAsync(modelId: String, autoSelected: Boolean) {}

    /**
     * Called when upgrade detection finds a model upgrade between what's installed
     * and what the registry now advertises. Fires once per detected upgrade.
     */
    suspend fun onUpgradeAvailableAsync(upgrade: UpgradeInfo) {}

    /**
     * (RT-04) Called when the runtime hot-swaps from one model in the fallback
     * chain to the next under memory pressure.
     */
    suspend fun onBrownoutAsync(from: String, to: String, reason: BrownoutReason) {}
}

/** Default no-op observer — hosts can subclass and override what they need. */
open class AIObserverBase : IAIObserver

// ---------------------------------------------------------------------------
// AIOptions (AIOptions.cs)
// ---------------------------------------------------------------------------

/**
 * Configuration for [AIService] and the loopback transport. All fields have safe
 * defaults so callers can `AIOptions()` and get a working instance. Mirrors C#
 * `AIOptions`.
 */
data class AIOptions(
    // ── Model ──
    val modelId: String? = null,
    val modelPath: String? = null,

    // ── Inference ──
    val systemPrompt: String = "You are B!, a helpful on-device assistant.",
    val defaultGenerationOptions: GenerationOptions? = null,
    val contextSize: Int? = null,
    val threadCount: Int? = null,
    val warmOnStart: Boolean = true,

    // ── Tools ──
    val toolBridge: IToolBridge? = null,

    // ── Observers ──
    val observer: IAIObserver? = null,

    // ── v2.0 Sensorium ──
    val deviceContext: IDeviceContext? = null,
    val catalogClient: ModelScopeCatalogClient? = null,
    val requiredCapabilities: Int = ChatCapability.DEFAULT,
    val checkForUpgradesOnStart: Boolean = false,
    val modelStorageDirectory: String? = null,

    // ── v2.0 Memory / RAG ──
    val episodicMemory: IEpisodicMemoryStore? = null,
    val ragTopK: Int = 5,

    // ── v2.0 Persona evolution ──
    val personaStore: IPersonaStore? = null,
    val personaUserId: String = "default",

    // ── v2.0 Feedback signals ──
    val feedbackStore: IFeedbackStore? = null,

    // ── v2.0 Agentic loop ──
    val agenticMaxIterations: Int? = null,

    // ── HttpLoopbackEndpoint configuration ──
    val loopbackPort: Int = 0,
    val loopbackToken: String? = null,

    // ── v2.1 Native runtime ──
    val nativeLibDir: String? = null,

    // ── v2.1 Model management ──
    val modelStorageDir: String = "models",
    val wifiOnlyModelDownload: Boolean = true,

    // ── v2.1 Cloud fallback ──
    val cloudFallbackEnabled: Boolean = false,
    val cloudFallbackEndpoint: String? = null,
    val cloudFallbackToken: String? = null,
    val cloudFallbackRamThresholdBytes: Long = 2L * 1024 * 1024 * 1024,

    // ── v2.1 Thermal management ──
    val thermalPauseEnabled: Boolean = true,
    val thermalService: IThermalThrottleService? = null,

    // ── v3.0 Scheduled tasks ──
    val scheduledTaskStore: IScheduledTaskStore? = null,

    // ── v3.0 Affect model ──
    val affectStore: IAffectStore? = null,

    // ── v3.0 Goal tracking ──
    val goalStore: IGoalStore? = null,
) {
    companion object {
        /**
         * Generates a cryptographically random 32-byte token, base64-encoded.
         * Used by [com.bhengubv.circleai.hosting.HttpLoopbackEndpoint] when
         * [loopbackToken] is null. Mirrors C# `AIOptions.GenerateRandomToken`.
         */
        fun generateRandomToken(): String {
            val bytes = ByteArray(32)
            SecureRandom().nextBytes(bytes)
            return Base64.getEncoder().encodeToString(bytes)
        }
    }
}
